design/baremetal-operator/bmh_live_boot.md

```diff
@@ -0,0 +1,121 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Add boot-iso API to BareMetalHost
+
+## Status
+
+implementable
+
+## Summary
+
+Add a new interface so that it is possible to boot arbitrary
+iso images via Ironic.
+
+## Motivation
+
+In some circumstances it is desirable to boot an existing iso image
+instead of having Ironic boot IPA, for example:
+
+To reduce boot time for ephemeral workloads, it may be possible to boot an iso
+and not deploy any image to disk (saving the time to write the image and reboot)
+
+Where an alternative installer exists that is distributed as a live-iso, it
+may be desirable to leverage that toolchain instead of the IPA deploy ramdisk,
+for example the [fedora-coreos installer](https://docs.fedoraproject.org/en-US/fedora-coreos/bare-metal/)
+
+An [Ironic API](https://docs.openstack.org/ironic/latest/admin/ramdisk-boot.html)
+exists that supports this but it's not currently accessible via metal3.
+
+### Goals
+
+Expose the Ironic API to boot iso images via metal3.
+
+### Non-Goals
+
+This only considers the ability to boot an iso image (which Ironic supports
+via both redfish-virtualmedia and iPXE), not live kernel/ramdisk images.
+
+Ironic does support booting kernel/ramdisk images though, so the interface
+we decide on shouldn't preclude adding that capability in future.
+
+## Proposal
+
+Add a new field to the BareMetalHost spec which indicates that instead of
+deploying an image to disk, a live-iso image will be booted:
+
+```yaml
+  spec:
+    liveImage:
+      url: http://1.2.3.4/image.iso
+    online: true
+```
+
+Note that in this mode, `rootDeviceHints` and `userData` will not be used
+since the image won't be written to disk, and Ironic doesn't
+support passing user-data in addition to the iso attachment at present.
+
+At some future point it would be desirable for Ironic to add the ability
+to pass userData - then a generic iso could be booted with the ability
+to define customization via the userData field in the normal way.
+
+## Design Details
+
+When this mode is selected, we need to configure Ironic to use the [ramdisk
+deploy interface](https://docs.openstack.org/ironic/latest/admin/ramdisk-boot.html)
+which is not currently enabled in ironic-image.
+
+We will also need to write the provided ISO URL into the
+[instance_info boot_iso field](https://docs.openstack.org/ironic/latest/admin/drivers/redfish.html#virtual-media-ramdisk)
+
+### Risks and Mitigations
+
+- Currently there is no detachment API in ironic, so in the case where an installer
+  iso is booted, we don't have a way to detach the virtualmedia and will have to
+  rely e.g on efibootmgr in whatever iso gets booted to ensure the correct boot device.
+
+- The inspection of a BMH will still rely on booting IPA and there's no support for
```

Comment:
We wouldn't do it as part of this enhancement, but should we add a flag to disable inspection to support workflows that get the hardware details another way?

Reply:
Yes we could do, I was thinking we'd rely on the existing interface e.g pass some status annotation at the point of creating the BMH, but then we'd need to change behavior so it can be updated after the hardware details are collected.  

A flag may be clearer though, I can either include that here or in a follow-up proposal when we have the boot-iso part working?

Reply:
For the use case you're working on, the host resource needs to exist before the ISO with the alternative agent is booted, so there's no opportunity to collect valid data in advance. We could pass fake data, but that feels wrong.

How about if we say that by default if there is a live image on the host we would not perform inspection, under the assumption that we either don't care about hardware details in those cases or the details will come from somewhere else later. Then we can use other design docs to cover an API for setting hardware details after the first reconciliation and to add a flag like `spec.liveImage.enableInspection` to restore the ability to (optionally) perform inspection with IPA before attaching the ISO.

Reply:
I'm not sure making inspection conditional on the spec.liveImage will work, because we may create the BMH resource then set the image reference later, and in that case we want the BMH to enroll and be ready to boot the iso, but not perform the existing inspection steps?

Perhaps adding an explicit flag elsewhere in the spec would be clearer?

Reply:
Let's just have an annotation that causes inspection to be skipped. That's simple to do, works for all use cases, and avoids making it part of the Spec which is a bit icky.
There is a proposal I can't find right now to add an annotation to force a new inspection. Maybe we could use the same key with different values (e.g. `never` to skip it even the first time vs. `refresh` to trigger a new one).

Note that to add the data externally later on you'll need RBAC write access to the Status, which is unusual.

Reply:
@zaneb I think this deserves a separate enhancement proposal - I suspect we need a way to skip inspection, but also add the data via the existing status annotation (avoiding the status RBAC issue you mention), which beyond the scope of this PR - I'll start a new one if that sounds reasonable.

Reply:
I agree with moving the inspection discussion to another enhancement.